crypto/provider_core.c

```diff
@@ -432,8 +460,61 @@ static int provider_activate(OSSL_PROVIDER *prov)
             prov->query_operation =
                 OSSL_get_provider_query_operation(provider_dispatch);
             break;
+#ifndef OPENSSL_NO_ERR
+# ifndef FIPS_MODE
+        case OSSL_FUNC_PROVIDER_GET_REASON_STRINGS:
+            p_get_reason_strings =
+                OSSL_get_provider_get_reason_strings(provider_dispatch);
+            break;
+# endif
+#endif
+        }
+    }
+
+#ifndef OPENSSL_NO_ERR
+# ifndef FIPS_MODE
+    if (p_get_reason_strings != NULL) {
+        const OSSL_ITEM *reasonstrings = p_get_reason_strings(prov->provctx);
+        size_t cnt, cnt2;
+
+        /*
+         * ERR_load_strings() handles ERR_STRING_DATA rather than OSSL_ITEM,
+         * although they are essentially the same type.
+         * Furthermore, ERR_load_strings() patches the array's error number
+         * with the error library number, so we need to make a copy of that
+         * array either way.
+         */
+        cnt = 1;                 /* One for the terminating item */
+        while (reasonstrings[cnt].id != 0)
+            cnt++;
+
+        /* Allocate one extra item for the "library" name */
+        prov->error_strings =
+            OPENSSL_zalloc(sizeof(ERR_STRING_DATA) * (cnt + 1));
+        if (prov->error_strings == NULL)
+            return 0;
+
+        /*
+         * We copy reasonstrings item 0..cnt-1 to prov->error_trings
+         * positions 1..cnt.  To do that effectively, we decrement cnt
+         * but save it's previous value, and getting a correct mapping
+         * of indexes.
+         */
+        while ((cnt2 = cnt--) > 0) {
```

Comment:
I think this code is more clear:
```
prov->error_strings[cnt2].error = ERR_PACK(prov->error_lib, 0, 0);
prov->error_strings[cnt2].string = prov->name;
for (cnt2 =1 ; cnt2 < cnt; cnt2++) {
...
}

```

Reply:
Sure, I can do that

Reply:
Better?

Reply:
Well I'd do zero based and do -1 on the LHS instead of +1 on the RHS side.

JUST KIDDING.

thanks.

Reply:
> JUST KIDDING.

[pfeww] :wink: